src/libNode/Node.cpp

```diff
@@ -271,43 +228,254 @@ void Node::Prepare(bool runInitializeGenesisBlocks) {
+  }
 
-  bool tx_result;
-  std::thread tTx(&Retriever::RetrieveTxBlocks, m_retriever.get(),
-                  std::ref(tx_result));
-  // retriever->RetrieveTxBlocks(tx_result);
+  BlockStorage::GetBlockStorage().GetDSCommittee(
+      m_mediator.m_DSCommittee, m_mediator.m_ds->m_consensusLeaderID);
+
+  std::vector<unsigned char> metaRes;
+  if (BlockStorage::GetBlockStorage().GetMetadata(MetaType::WAKEUPFORUPGRADE,
+                                                  metaRes)) {
+    if (metaRes[0] == '1') {
+      wakeupForUpgrade = true;
+      BlockStorage::GetBlockStorage().PutMetadata(MetaType::WAKEUPFORUPGRADE,
+                                                  {'0'});
+    }
+  }
 
+  /// Retrieve Tx blocks, relative final-block state-delta from persistence
   bool st_result = m_retriever->RetrieveStates();
+  bool tx_result;
+  m_retriever->RetrieveTxBlocks(tx_result, wakeupForUpgrade);
+
+  if (!tx_result) {
+    return false;
+  }
+
+  /// Retrieve lacked Tx blocks from lookup nodes
+  uint64_t oldTxNum = m_mediator.m_txBlockChain.GetBlockCount();
+  m_synchronizer.FetchLatestTxBlocks(m_mediator.m_lookup,
+                                     m_mediator.m_txBlockChain.GetBlockCount());
+  LOG_GENERAL(INFO, "Retrieve final block from lookup node, please wait...");
 
-  tDS.join();
-  tTx.join();
+  unique_lock<mutex> lock(m_mediator.m_lookup->m_MutexCVSetTxBlockFromSeed);
+  if (m_mediator.m_lookup->cv_setTxBlockFromSeed.wait_for(
+          lock, chrono::seconds(RECOVERY_SYNC_TIMEOUT)) == cv_status::timeout) {
+    LOG_GENERAL(WARNING,
+                "GetTxBlockFromLookupNodes Timeout, cannot get final block "
+                "from lookup node!");
+  }
+
+  if (m_mediator.m_txBlockChain.GetBlockCount() > oldTxNum + 1) {
+    LOG_GENERAL(INFO, "Node recovery too late, apply re-join process instead");
+    return false;
+  }
+
+  /// Retrieve lacked final-block state-delta from lookup nodes
+  if (m_mediator.m_txBlockChain.GetBlockCount() > oldTxNum) {
+    m_mediator.m_lookup->GetStateDeltaFromLookupNodes(
+        m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum());
+    LOG_GENERAL(
+        INFO,
+        "Retrieve final block state delta from lookup node, please wait...");
+    unique_lock<mutex> lock(
+        m_mediator.m_lookup->m_MutexCVSetStateDeltaFromSeed);
+    if (m_mediator.m_lookup->cv_setStateDeltaFromSeed.wait_for(
+            lock, chrono::seconds(RECOVERY_SYNC_TIMEOUT)) ==
+        cv_status::timeout) {
+      LOG_GENERAL(WARNING,
+                  "GetStateDeltaFromLookupNodes Timeout, cannot get final "
+                  "block state delta from lookup node!");
+    }
+  }
+
+  /// Save coin base for final block, from last DS epoch to current TX epoch
+  for (uint64_t blockNum =
+           m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() -
+           (m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() %
+            NUM_FINAL_BLOCK_PER_POW);
```

Comment:
If a fallback happened, which will result a ds block being generated at any epoch, this starting `blockNum` may not fit. Alternatively, You can use the epoch number obtained from DSBlockHeader after #788 was merged

Reply:
ok, will do that then, in a separated PR